### DISCUSSION TITLE
Problems with gyro and turning

### DISCUSSION THREAD

**Discussion by SnakeJuice** [pybricks/support#1289] [Q&A] [answered]
Currently I'm working with a LEGO Mindstorm EV3 and using a gyro module for turning and going straight. 

It manages to run the first_run() function normally then when it reaches second_run() and inside this function it runs the function gyro_left_run() it starts to spin forever to the left. I don't know what is wrong, I've checked the documentation and everything. 

This are all the functions I have in my project. 

```python
def gyro_reset():
    gyro.reset_angle(0)
    while True:
        if gyro.angle() == 0:
            break
    while True:
        if gyro.speed() == 0:
            break


def gyro_straight(distance):
    target = gyro.angle()
    gain = 1.5 # Gain
    robot.reset() # Reset robot
    while robot.distance() < distance:
        correction = target - gyro.angle()
        turn_power = correction * gain
        robot.drive(100, turn_power)
        print("Robot distance:", robot.distance(), "Target:", distance)
        print("Gyro angle:", gyro.angle(), "Target:", target)
    robot.stop()
    

def gyro_right_turn(degrees):
    gyro_reset() # Reset gyro
    while gyro.angle() < degrees:
        left_motor.run(100)
    left_motor.stop()


def gyro_left_turn(degrees):
    gyro_reset() # Reset gyro
    while gyro.angle() < degrees:
        right_motor.run(100)
    right_motor.stop()
    

def first_run():
    print("first run start")
    gyro_straight(1200)
    gyro_right_turn(90)
    gyro_straight(20)
    gyro_right_turn(90)
    print("1")


def second_run():
    print("second run start")
    gyro_straight(1200)
    gyro_left_turn(90)
    gyro_straight(20)
    gyro_left_turn(90)
    print("2")

first_run()
second_run()
first_run()
second_run()
```

**Comment by laurensvalk** [comment, marked as answer]
```python
def gyro_reset():
    gyro.reset_angle(0)
    while True:
        if gyro.angle() == 0:
            break
    while True:
        if gyro.speed() == 0:
            break
```

If your goal is to reset the angle value, just calling `gyro.reset_angle(0)` is sufficient.

I'm not sure the other parts of this function are needed. In fact, if the robot is turning while you call this, `gyro.speed()` is nonzero, so it is always going to be stuck in the second loop. And thus spin forever which I think you might be seeing.

**Reply by SnakeJuice** [reply]
Thank you. I think that was the problem. Also, changed the logic of the turns because I was doing it wrong.

This is the revised version if someone by any chance needs it

```python
# Gyro turn left
def gyro_right_turn(degrees):
    gyro_reset()
    target_angle = gyro.angle() + degrees

    while gyro.angle() < target_angle:
        left_motor.run(100)
    left_motor.stop()

#Gyro turn right
def gyro_left_turn(degrees):
    gyro_reset() # Reset gyro
    target_angle = gyro.angle() - degrees

    while gyro.angle() > target_angle:
        right_motor.run(100)
    right_motor.stop()
```

**Comment by FLL-Team-24277** [comment]
I am curious as to why you are not using the pybricks robot.turn() function. Do you think there is some advantage to coding your own turn functions as you have here?

**Reply by laurensvalk** [reply]
The EV3 does not have a built-in gyro like the Spike hubs do. The external gyro doesn't give us as much control, so automatic integration with `turn` is not included there.
